### ISSUE TITLE
cmd/go: diamond CGO dependency produces a symbol collision

### ISSUE THREAD

**Issue by colinmarc** [golang/go#16463]
Please answer these questions before submitting your issue. Thanks!
1. What version of Go are you using (`go version`)?
   
   ```
   go version go1.6.3 darwin/amd64
   ```
2. What operating system and processor architecture are you using (`go env`)?
   
   ```
   GOARCH="amd64"
   GOBIN=""
   GOEXE=""
   GOHOSTARCH="amd64"
   GOHOSTOS="darwin"
   GOOS="darwin"
   GOPATH="/Users/colin/golang"
   GORACE=""
   GOROOT="/usr/local/go"
   GOTOOLDIR="/usr/local/go/pkg/tool/darwin_amd64"
   GO15VENDOREXPERIMENT="1"
   CC="clang"
   GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fno-common"
   CXX="clang++"
   CGO_ENABLED="1"
   ```
3. What did you do?
   - Clone https://github.com/stripe/sequins
   - Check out https://github.com/stripe/sequins/commit/011971aa3b593f169f8f53a7e07d427377f4bcb8 (the `compressed-sequencefiles` branch)
   - Run `make sequins`.
4. What did you expect to see?
   
   Successful build output!
5. What did you see instead?
   
   It's failing to build on the last step with duplicate symbols:
   
   ```
     duplicate symbol _snappy_compress in:
         /var/folders/3d/ymqb8kcx5jv4s2gsqy7rd83h0000gn/T/go-link-410495653/000000.o
         /var/folders/3d/ymqb8kcx5jv4s2gsqy7rd83h0000gn/T/go-link-410495653/000003.o
   ```

I can also reproduce on linux; here's a [travis build](https://api.travis-ci.org/jobs/146532902/log.txt?deansi=true).

The issue seems to be that I have a diamond dependency on a local static library. I vendor and build a version of snappy, which is in turn used by [sparkey](https://github.com/spotify/sparkey), which I also build, and that is in turn used by a vendored version of [go-sparkey](https://github.com/bsm/go-sparkey), which is used in my app.

That had all been working fine (you can check out the `1.0` branch to see it in action). The issue arose when I introduced another dependency on snappy, in [`sequencefile/snappy.go`](https://github.com/stripe/sequins/blob/011971aa3b593f169f8f53a7e07d427377f4bcb8/sequencefile/snappy.go). Then the build started breaking.

To summarize, the dependency graph looks like this (see the [Makefile](https://github.com/stripe/sequins/blob/011971aa3b593f169f8f53a7e07d427377f4bcb8/Makefile) for reference):

![](https://cloud.githubusercontent.com/assets/821775/17043715/4c4c306e-4f6e-11e6-93b1-e88db28a87b6.png)

Everything was working before I added the brach on the right; the [linked commit](https://github.com/stripe/sequins/commit/011971aa3b593f169f8f53a7e07d427377f4bcb8) contains the full set of changes that break the build.

Thanks in advance for the help! Sorry this is gnarly.

**Comment by rsc** [comment]
I think the problem is the CGO_LDFLAGS being set, as seen in the Travis log:

	CGO_LDFLAGS="/home/travis/gopath/src/github.com/stripe/sequins/build/lib/libsparkey.a /home/travis/gopath/src/github.com/stripe/sequins/build/lib/libsnappy.a /home/travis/gopath/src/github.com/stripe/sequins/build/lib/libzookeeper_mt.a -lrt -lm -lstdc++"

One step of the cgo build is to take the .o files that have been generated so far for the package (containing the Go-specific wrappers) and merge them into one by using (I am simplifying a little):

	$CC -o _all.o *.o $CGO_LDFLAGS

The $CGO_LDFLAGS in this invocation has -l arguments stripped out, precisely to avoid this problem, but not explicit paths like foo.a. I believe that if you change your CGO_LDFLAGS to:

	CGO_LDFLAGS="-L/home/travis/gopath/src/github.com/stripe/sequins/build/lib -lsparkey -lsnappy -lzookeeper_mt -lrt -lm -lstdc++"

then your build should work. We know to remove *.dylib, *.so, and *.dll from the CGO_LDFLAGS. Perhaps we should also remove *.a.

@ianlancetaylor, what do you think about removing *.a in src/cmd/go/build.go's builder.collect?

**Comment by ianlancetaylor** [comment]
I removing .o and .a files from `CGO_LDFLAGS` in `builder.collect` makes sense.
